lib/teiserver/account/libs/user_totp_lib.ex

```diff
@@ -0,0 +1,237 @@
+            {:error, changeset}
+        end
+
+      {:inactive, _} ->
+        :ok
+    end
+  end
+
+  defp increase_wrong_otp_counter(%TOTP{} = totp) do
+    from(t in TOTP, where: t.user_id == ^totp.user_id)
+    |> Repo.update_all(inc: [wrong_otp: 1])
+  end
+
+  def reset_wrong_otp_counter(totp) do
+    from(t in TOTP, where: t.user_id == ^totp.user_id)
+    |> Repo.update_all(set: [wrong_otp: 0])
+  end
+
+  def validate_totp(user_or_secret, otp, time \\ System.os_time(:second))
+
+  @spec validate_totp(User.t(), String.t(), integer) ::
+          {:ok, :valid | :grace} | {:error, :inactive | :invalid | :used | :locked}
+  def validate_totp(%User{id: user_id}, otp, time) do
+    {status, totp} = get_user_totp(user_id)
+
+    with :active <- status,
+         false <- get_account_locked(user_id),
+         {:ok, info} <- validate_totp(totp.secret, otp, time, since: totp.last_used) do
+      case info do
+        :valid ->
+          set_last_used(user_id, time)
+
+        :grace ->
+          set_last_used(user_id, time - @grace)
+      end
+
+      reset_wrong_otp_counter(totp)
+      {:ok, info}
+    else
+      :inactive ->
+        {:error, :inactive}
+
+      true ->
+        {:error, :locked}
+
+      {:error, status} ->
+        if status == :invalid do
+          increase_wrong_otp_counter(totp)
+        else
+          reset_wrong_otp_counter(totp)
+        end
+
+        {:error, status}
+    end
+  end
+
+  @spec validate_totp(binary, String.t(), integer) :: {:ok, :valid | :grace} | {:error, :invalid}
+  def validate_totp(secret, otp, time) do
+    cond do
+      NimbleTOTP.valid?(secret, otp, time: time) ->
+        {:ok, :valid}
+
+      NimbleTOTP.valid?(secret, otp, time: time - @grace) ->
+        {:ok, :grace}
+
+      true ->
+        {:error, :invalid}
+    end
+  end
+
+  @spec validate_totp(binary, String.t(), integer, keyword()) ::
+          {:ok, :valid | :grace} | {:error, :invalid}
+  defp validate_totp(secret, otp, time, since: nil) do
+    validate_totp(secret, otp, time)
+  end
+
+  @spec validate_totp(binary, String.t(), integer, keyword()) ::
+          {:ok, :valid | :grace} | {:error, :invalid | :used}
+  defp validate_totp(secret, otp, time, since: last_used) do
+    cond do
+      NimbleTOTP.valid?(secret, otp, time: time, since: last_used) ->
+        {:ok, :valid}
+
+      NimbleTOTP.valid?(secret, otp, time: time - @grace, since: last_used) ->
+        {:ok, :grace}
+
+      true ->
+        # Second test is needed. If :since is provided and NimbleTOTP.valid? returns false, it could either be that the OTP is wrong, or that it got used.
+        # To figure out which one it is, we need to test a second time without since, as a false this time indicates that the OTP is invalid, and a True that it got used
+        case validate_totp(secret, otp, time) do
+          {:error, :invalid} ->
+            {:error, :invalid}
+
+          _ ->
+            {:error, :used}
+        end
+    end
```

Comment:
This function is a duplicate of `validate_totp/3`. You should instead the first function set a default for the `since` argument and then delegate to this one. Something like

```elixir
defp validate_totp(secret, otp, time), do: validate_totp(secret, otp, time, since: nil)`
```
and adjust that function to handle a `since: nil`
This way you don't have the duplicate logic.